docs/proposals/service-binding-specification/implementing-specification.md

```diff
@@ -0,0 +1,129 @@
+Operator employs such as how the application is defined in the CRD, how a binding secret is
+projected in an application container and etc. The [Service Binding feature comparison document](./service-binding-feature-comparison.pdf) lists the differences between the two.
+
+One of the major gaps between the CRDs defined by the two approaches is that the Service Binding
+Operator supports defining multiple services within a CR. Whereas the specification allows defining
+one service per CR. Here are two main reasons why the specification is in favor of defining one
+service per CR:
+
+- It is easier for tools created around the service binding to add and remove services by simply
+  adding or removing independent CRs instead of having to deal with editing an array of services
+  within a CR. In addition, editing the service array in a CR might cause CR to be corrupted unless
+  concurrent updates are well synchronized across a tool or tools. However, by defining a single
+  service per CR, adding or removing services simply becomes `kubectl apply ...` and
+  `kubectl delete ...`.
+
+- Defining one service per CR allows a cleaner association of attributes with the service. For example,\
+  when we identify a set of extra mappings or environment variables in a `ServiceBinding`, which
+  service does that get associated with if we have multiple services? Or if we want to change the
+  service name or type that gets mounted, we again run into the issue that they all require a way
+  to identify the service you want to apply those attributes against.
+
+Therefore, allowing users of the Service Binding Operator to choose between the specification and the
+current mechanism is likely to increase the user base of the operator, having a more vibrant community
+of service providers.
+
+### Goals
+
+Goals for this proposal are listed as follows:
+
+- Allow users of the Service Binding Operator to have a clear separation in choosing between the
+specification and the current mechanism for binding applications to services.
+
+### Non-Goals
+
+- Implementation details about how the specification should be implemented.
+
+## Proposal
+
+To avoid tangling the implementation between the current codebase and the codebase to support the
+specification, this proposal suggests creating two different channels for the Service Binding
+Operator (e.g. `current` and `spec`) to avoid mixing of implementation details between the two
+codebases. This way users can freely choose what set of resources they expect the operator to act on at
+the operator install time.
+
+### User Stories
+
+#### Story 1
+
+When installing the Service Binding Operator, a user can choose which of the two channels
+to install the Service Binding Operator from. Depending on the channel, the operator installs different
+sets of Custom Resource Definitions (CRDs) and would watch for different sets of resources in the
+cluster.
+
+### Implementation Details/Notes/Constraints
```

Comment:
As constraints/notes could we outline 
* Which use cases will be supported as is ( but with a different API ) and 
* Which ones the new implementation isn't planning on supporting.

This [doc](https://github.com/redhat-developer/service-binding-operator/blob/d82c45ffbe6cf982fb3bc5c35f0be7e6f241066a/docs/proposals/service-binding-specification/feature-comparison.pdf) this does a great job of API comparison, though.

Some questions I would have is:

1. How do we handle custom binding information ( known as customEnvVar here) which is constructed from the content in the spec/status and/or referenced secret/configmap ? ( no annotations/descriptors ) 

2. Would the new implementation have a way to auto-detect binding resources ( `detectBindingResources` ) ?

3. Multiple service resource use case: If a connection string has to be constructed from two resources namely, "DBInstance" and "DBCredentials" , how does a user do that?

Reply:
hey @sbose78 - here are some thoughts on the questions:
1. The spec allows for the creation of custom mappings using hardcoded Strings + values available from the `binding Secret`.  So the scope of available data for mapping is the information the provisioned service has made available - not its spec/status.  An example [here](https://github.com/k8s-service-bindings/spec/blob/master/README.md#mappings-example-resource).   This custom mapping is then added to your available binding Secret.   For security reasons we volume mount all of the binding Secret's keys, but you can choose which ones you want to make available as env vars, sample [here](https://github.com/k8s-service-bindings/spec/blob/master/README.md#environment-variables-example-resource).

2.  This capability has been discussed in the spec and deemed that can be a useful value-add of implementations outside of the spec.  

3.  A user would have one `ServiceBinding` per binding resource, and could then choose how to concatenate the two inside the app - the choices would be via volume mount or env var.   Although I would add that this scenario hints at an opportunity for improving the provisioned service:  the information needed to connect should be available in a single resource (e.g. the `DBCredentials` should contain the information about which DB URL the credentials apply, hence not needing a separate `DBInstance` resource for this scenario).

Reply:
Note, everything in Service Binding will be supported with volume mounts and env vars, mostly in a mutually exclusive way.

> The spec allows for the creation of custom mappings using hardcoded Strings + values available from the binding Secret

Note, this pattern is supported in a resource-agnostic way : 

```
apiVersion: operators.coreos.com/v1alpha1
kind: ServiceBinding
metadata:
  name: binding-request
  namespace: service-binding-demo
spec:
  application:
    name: java-app
    group: apps
    version: v1
    resource: deployments
  services:
  - group: ""
    version: v1
    kind: Secret
    name: any-secret
    id: postgresDB
  customEnvVar:
    - name: accountServiceUri
      value: ':{{ any-secret.data.key1:any-secret.data.key2 }}'
```
Are you proposing we drop support for spec/status/data-based introspection, or have both co-exist in the API?


> A user would have one ServiceBinding per binding resource, and could then choose how to concatenate the two inside the app 

That implies that the user has to
* Create 2 `ServiceBindings`
* Modify application code to stitch the information from individual secrets


> Although I would add that this scenario hints at an opportunity for improving the provisioned service: the information needed to connect should be available in a single resource (e.g. the DBCredentials should contain the information about which DB URL the credentials apply, hence not needing a separate DBInstance resource for this scenario)

This implies, that the service provider would have to make changes to the controller code, even though separating instance creation and credential creation is often a conscious design choice.

Did I understand you correctly ? :)

Reply:
> Are you proposing we drop support for spec/status/data-based introspection, or have both co-exist in the API?

I think this is a case where the functionality of SBO is a superset of what the spec is proposing, so we should be able to happily comply with the spec while keeping the value-add functionality.

> even though separating instance creation and credential creation is often a conscious design choice.

I would assert that this provider design choice forces consumers to deal with two CRs, so it's acceptable that the app would have these as separate values.  In many cases the app won't even be concatenating these values together into a single string, they would use them in separate places - e.g. adding the credentials in headers, and URL in the REST request.  I think this goes down the road of multiple vs single provisioned service references =)

Reply:
> I would assert that this provider design choice forces consumers to deal with two CRs, so it's acceptable that the app would have these as separate values. 

A conscious design choice by the provider doesn't necessarily mean the complexity has to be passed on to the app developer. The current ServiceBinding API implementation shows that consumer workloads *don't* have to deal with two different `ServiceBindings` if designed well.

One can make do with one `ServiceBinding` CR and construct amazing custom variables that could be used by both env vars & volume mounts.

Reply:
yup, that's true.  Although what I have seen is that the advantages of having separate ServiceBinding CRs outweigh the advantages of having single multi-service ServiceBinding CR (summarized [here](https://github.com/redhat-developer/service-binding-operator/pull/608#discussion_r482276288)).

Reply:
> it's easier for a tool to add and remove services by simply adding or removing independent CRs, instead of having to deal with editing (and potentially corrupting, if concurrent updates are not well synchronized) an array inside a CR. It simply becomes "kubectl apply .." and "kubectl delete .."

You can run into issues with concurrent updates to any CR's spec if the status is constantly changing - try updating a _just created_ `Deployment`. Also, supporting multiple services doesn't force any tool or anyone to also use multiple services - it can still refer to one service if it wants to ;)

> cleaner association of attributes with the service. For example, when we identify a set of extra mappings or environment variables in a ServiceBinding, which service does that get associated with if we have multiple services? Or if we want to change the service name or type that gets mounted, we again run into the issue that they all require a way to identify the service you want to apply those attributes against.

This is a solved problem - we already support this today in a very clean way using the group.version.kind.name notation or by simply using the service "id". You can use the "id" (or an equivalent identifier) in the mount path if you wish to.

Reply:
I think a big issue with using `id` to distinguish multiple services within a single CR is that every single attribute of your CR which impacts the projection now needs to be "id aware".  So it's the mappings, the env vars, the name of the projection, the status, etc.  Actually, the status of the CR is probably a whole separate topic of complexity, since now there would be a partial success cases where maybe some of the services were not yet ready, etc. 

In the spec we had a few hangouts where we discussed this, and arrived at the conclusion that a single service CR was also much simpler to work with, even when adding mappings and env vars: [sample](https://github.com/k8s-service-bindings/spec#environment-variables-example-resource)

Reply:
> I think a big issue with using id to distinguish multiple services within a single CR is that every single attribute of your CR which impacts the projection now needs to be "id aware". So it's the mappings, the env vars, the name of the projection, the status, etc. 

If a workload talks to multiple service resources, you have to _namespace_ things on the workload side  **anyway** :) Also, our code already has binding secret generation strategies that support this :) 

> a single service CR was also much simpler to work with

* Was it hard to implement ? Yes, there's work.
* Is it easy to use as a user ? Yes :) ( less code changes on the user's app, multiple Service Binding CRs **not** needed, will work with *any* backing service )

> the status of the CR is probably a whole separate topic of complexity, since now there would be a partial success cases where maybe some of the services were not yet ready, etc.

There's prior art we can take inspiration from https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate , again, it is a bit of work for the implementor, but do-able in a clean way to help the user.